components/brave_vpn/browser/connection/brave_vpn_region_data_manager.cc

```diff
@@ -0,0 +1,376 @@
+
+namespace brave_vpn {
+
+using ConnectionState = mojom::ConnectionState;
+
+namespace {
+
+base::Value::Dict GetValueFromRegion(const mojom::Region& region) {
+  base::Value::Dict region_dict;
+  region_dict.Set(kRegionContinentKey, region.continent);
+  region_dict.Set(kRegionNameKey, region.name);
+  region_dict.Set(kRegionNamePrettyKey, region.name_pretty);
+  region_dict.Set(kRegionCountryIsoCodeKey, region.country_iso_code);
+  return region_dict;
+}
+
+bool IsValidRegionValue(const base::Value::Dict& value) {
+  if (!value.FindString(kRegionContinentKey) ||
+      !value.FindString(kRegionNameKey) ||
+      !value.FindString(kRegionNamePrettyKey) ||
+      !value.FindString(kRegionCountryIsoCodeKey)) {
+    return false;
+  }
+
+  return true;
+}
+
+mojom::Region GetRegionFromValue(const base::Value::Dict& value) {
+  mojom::Region region;
+  if (auto* continent = value.FindString(brave_vpn::kRegionContinentKey)) {
+    region.continent = *continent;
+  }
+  if (auto* name = value.FindString(brave_vpn::kRegionNameKey)) {
+    region.name = *name;
+  }
+  if (auto* name_pretty = value.FindString(brave_vpn::kRegionNamePrettyKey)) {
+    region.name_pretty = *name_pretty;
+  }
+  if (auto* country_iso_code =
+          value.FindString(brave_vpn::kRegionCountryIsoCodeKey)) {
+    region.country_iso_code = *country_iso_code;
+  }
+  return region;
+}
+
+bool ValidateCachedRegionData(const base::Value::List& region_value) {
+  for (const auto& value : region_value) {
+    // Make sure cached one has all latest properties.
+    if (!value.is_dict() || !IsValidRegionValue(value.GetDict())) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+std::vector<mojom::Region> ParseRegionList(
+    const base::Value::List& region_list) {
+  std::vector<mojom::Region> regions;
+  for (const auto& value : region_list) {
+    DCHECK(value.is_dict());
+    if (!value.is_dict()) {
+      continue;
+    }
+    regions.push_back(GetRegionFromValue(value.GetDict()));
+  }
+
+  // Sort region list alphabetically
+  std::sort(regions.begin(), regions.end(),
+            [](mojom::Region& a, mojom::Region& b) {
+              return (a.name_pretty < b.name_pretty);
+            });
+  return regions;
+}
+
+}  // namespace
+
+BraveVPNRegionDataManager::BraveVPNRegionDataManager(
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory,
+    PrefService* local_prefs,
+    BraveVPNOSConnectionAPIBase* connection_api)
+    : url_loader_factory_(url_loader_factory),
+      local_prefs_(local_prefs),
+      connection_api_(connection_api) {
+  LoadCachedRegionData();
+}
+
+BraveVPNRegionDataManager::~BraveVPNRegionDataManager() = default;
+
+const std::vector<mojom::Region>& BraveVPNRegionDataManager::GetRegions()
+    const {
+  return regions_;
+}
+
+bool BraveVPNRegionDataManager::IsRegionDataReady() const {
+  return !regions_.empty();
+}
+
+void BraveVPNRegionDataManager::SetSelectedRegion(const std::string& name) {
```

Comment:
it is optional if possible would be great to move this method to connection api level. Seems like it allows us to remove connection api backwared dependency from this class. Using callback or interface maybe, wdyt?

Reply:
Good idea! will try in that day.

Reply:
@spylogsster Updated. PTAL again - https://github.com/brave/brave-core/pull/18009/commits/be64f0c62f764c28f054d1385b79e2a66413a0ef

Reply:
Looks awesome, thanks!